### DISCUSSION TITLE
Manually Triggered Source Generators

### DISCUSSION THREAD

**Discussion by Phylum123** [dotnet/roslyn#71067] [Ideas]
Would love to have the ability to manually trigger any source generator, perhaps add it to the context menu of the analyzer.  Also, have source generators that can only be manually triggered. This would allow for source generators to be used for anything that takes some time, like a SQL poco generation. Could easily add to the Generator attribute (bool isManualOnly).

**Comment by CyrusNajmabadi** [comment]
What is a SQL poco generator.  Can you give examples?

**Comment by Phylum123** [comment]
A source generator that creates classes (POCO - Plain old CLR objects) from SQL tables, views, etc. (ex: https://marketplace.visualstudio.com/items?itemName=SimonHughes.EntityFrameworkReversePOCOGenerator) They don't work if they are constantly being generated due to code change, due to the fact they access a database. You could also create a source generator that downloads and reads a swagger file and creates methods on a class to access a web api. There are many applications of a manually triggered source generator.

**Reply by CyrusNajmabadi** [reply]
Yeah, i would not use source-generators for these.  Source-generators are appropriate when you are analyzing C# code to generate code *that that same* c# code needs to use.  

In your case, you could accomplish this with just a trivial console tool that runs prior to the compiler running, which analyzes your SQL DBs or swagger files, or wahtever, and then just generates the proper C# files as the result.

This would, of course, then only run whenever you wanted it to.  SGs are for the incremental scenario where code has to bootstrap itself.  THat isn't the case here, so it's not a good tool for this job.

**Reply by Phylum123** [reply]
The reason for wanting to use a source generator instead of a console tool, is that the files can be locked, editted, etc. It would be nice to have generated code that is untouchable, and in memory, like when created with a source generator.
